gems/smithy-client/lib/smithy-client/endpoint_rules.rb

```diff
@@ -0,0 +1,188 @@
+# frozen_string_literal: true
+
+require 'cgi'
+require 'ipaddr'
+require 'uri'
+
+module Smithy
+  module Client
+    # Functions in the Smithy rules engine are named routines that
+    # operate on a finite set of specified inputs, returning an output.
+    # The rules engine has a set of included functions that can be
+    # invoked without additional dependencies, called the standard library.
+    module EndpointRules
+      # Regex that extracts anything in square brackets
+      BRACKET_REGEX = /\[(.*?)\]/
+
+      # An Endpoint resolved by an EndpointProvider
+      class Endpoint
+        # @param [String] :uri
+        # @param [Array<AuthScheme>] :auth_schemes ([])
+        # @param [Hash] :headers ({})
+        def initialize(uri:, properties: {}, headers: {})
+          @uri = uri
+          @properties = properties
+          @headers = headers
+        end
+
+        # The URI of the endpoint.
+        # @return [String]
+        attr_accessor :uri
+
+        # The authentication schemes supported by the endpoint.
+        # @return [Hash]
+        attr_accessor :properties
+
+        # The headers to include in requests to the endpoint.
+        # @return [Hash]
+        attr_accessor :headers
+      end
+
+      # Evaluates whether the input string is a compliant RFC 1123 host segment.
+      # When allowSubDomains is true, evaluates whether the input string is
+      # composed of values that are each compliant RFC 1123 host segments
+      # joined by dot (.) characters.
+      # @api private
+      # rubocop:disable Style/OptionalBooleanParameter
```

Comment:
I don't remember if we were able to fix this, but we should just solve this one if possible.

Reply:
I removed the default value for this.  It makes testing a little easier but all usage of it in endpoint rules requires the parameter, so it isn't needed.